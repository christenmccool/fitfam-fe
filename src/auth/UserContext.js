import React from 'react';

/** Context: provide access to user object throughout app */

const UserContext = React.createContext();

export default UserContext;
