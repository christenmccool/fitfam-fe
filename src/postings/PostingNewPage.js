import React, {useState, useEffect, useContext} from 'react';
import {useSearchParams, Link as RouterLink} from 'react-router-dom';
import moment from 'moment';

import Container from '@mui/material/Container';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import TextField from '@mui/material/TextField';
import MenuItem from '@mui/material/MenuItem';
import Button from '@mui/material/Button';


import FitFamApi from '../api/api';
import UserContext from '../auth/UserContext';
import SelectDate from '../common/SelectDate';
import FamilySelect from './FamilySelect';
import CustomPostingForm from './CustomPostingForm';
import AddWoPostingForm from './AddWoPostingForm';

/** Shows form for posting a workout to a family
 * Can either create a custom workout or add one from the database
 * 
 * PostingNewPage -> {CustomPostingForm, AddWoPostingForm}
 */
const PostingNewPage = () => {
  const {user, currFamId, setCurrFamId} = useContext(UserContext);
  const [searchParams, setSearchParams] = useSearchParams("");

  const initialDate = searchParams.get('date') || moment().format("YYYY-MM-DD");
  const [date, setDate] = useState(initialDate);
  const [familyId, setFamilyId] = useState(currFamId);

  const options = [
    {
      value: "search",
      label: "Search for workout"
    },
    {
     value: "create",
     label: "Create new workout"
    }
  ]
  const [option, setOption] = useState("create");

  const handleOptionChange =  async (event) => {
    const {value} = event.target;
    setOption(value);
  }

  useEffect(() => {
    const newDate = searchParams.get('date');
    if (newDate) {
      setDate(newDate);
    } 
  }, [searchParams, setSearchParams])

  function changeFamilyId(famId) {
    setFamilyId(famId);
    setCurrFamId(famId);
  }

  async function postNewWorkout(data) {
    try {
      const workout = await FitFamApi.createWorkout({name: data.woName, description: data.woDescription, scoreType: "Other / Text", createBy: user.id});
      const posting = await FitFamApi.createPosting(workout.id, familyId, date, user.id);
      return {success: true, postId: posting.id}
    } catch (err) {
      console.log(err);
      return {success: true, err}
    }
  }

  async function postExistingWorkout(workoutId) {
    try {
      const posting = await FitFamApi.createPosting(workoutId, familyId, date, user.id);
      return {success: true, postId: posting.id}
    } catch (err) {
      console.log(err);
      return {success: true, err}
    }
  }

  return (
    <Container align="center" maxWidth="md" sx={{backgroundColor: "#FFF", borderRadius: '10px'}}>
      <Box m={5} p={3}>
        <Box sx={{display: 'flex', justifyContent: 'center', alignItems: 'center', flexDirection: {xs: "column", sm: "row"} }}>
            <Box p={1}>
              < SelectDate />
            </Box>
            <Box p={1}>
              < FamilySelect 
                families={user.families}
                familyId={familyId}
                changeFamilyId={changeFamilyId}
              />
            </Box>
        </Box>

        <Grid item xs={12} sm={8} mt={1} mb={5} p={2}>
          <TextField
            fullWidth
            id="option"
            select
            label="Select option"
            color="primary"
            value={option}
            onChange={handleOptionChange}
            InputProps={{style: {fontSize: '20px', color: '#3f50b5'}}}
          >
            {options.map((option) => (
              <MenuItem key={option.value} value={option.value}>
                {option.label}
              </MenuItem>
            ))}
          </TextField>
        </Grid>

        {option === "create" ?
          <Box mt={3}>
            <CustomPostingForm 
              formType="create"
              date={date}
              submitPostForm={postNewWorkout}
            />
          </Box>
          :
          <Box mt={3}>
            <AddWoPostingForm 
              date={date}
              postExistingWorkout={postExistingWorkout}
            />
          </Box>
        }

        <Grid container justifyContent="flex-end">
          <Grid item>
            <Button
              component={RouterLink}
              to={`/?date=${date}`}
              type="button"
              size="large"
            >
              <Typography color="text.secondary">
                Cancel
              </Typography>
            </Button>
          </Grid>
        </Grid>
      </Box>
    </Container>
  )
}

export default PostingNewPage;