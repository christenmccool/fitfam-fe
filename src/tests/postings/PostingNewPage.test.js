import React from "react";
import { render, screen } from "@testing-library/react";
import UserEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import AdapterMoment from '@mui/lab/AdapterMoment';
import LocalizationProvider from '@mui/lab/LocalizationProvider';

import {demoUser } from "../testutils";

import PostingNewPage from "../../postings/PostingNewPage";
import UserContext from "../../auth/UserContext";


it('renders without crashing', function () {
  render(
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <UserContext.Provider value={{user: demoUser, currFamId: 1}}>
          <PostingNewPage />
        </UserContext.Provider>
      </LocalizationProvider>
    </MemoryRouter>
  );
});


it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <UserContext.Provider value={{user: demoUser, currFamId: 1}}>
          <PostingNewPage />
        </UserContext.Provider>
      </LocalizationProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});


it("renders posting add form", function () {
  render(
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <UserContext.Provider value={{user: demoUser, currFamId: 1}}>
          <PostingNewPage />
        </UserContext.Provider>
      </LocalizationProvider>
    </MemoryRouter>
  );
  expect(screen.getByLabelText("Create new workout")).toBeInTheDocument();  
  expect(screen.queryAllByText("Workout Name")[0]).toBeInTheDocument();  
  expect(screen.queryAllByText("Workout Description")[0]).toBeInTheDocument();  
  expect(screen.getByRole('button', {name: /submit/i})).toBeInTheDocument();
});


it("responds to user input", function () {
  render(
    <MemoryRouter>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <UserContext.Provider value={{user: demoUser, currFamId: 1}}>
          <PostingNewPage />
        </UserContext.Provider>
      </LocalizationProvider>
    </MemoryRouter>
  );

  const optionSelect = screen.getByRole('button', {name: /Select option/i});
  UserEvent.click(optionSelect);
  UserEvent.click(screen.getByRole('option', {name: /Create new workout/i}));
  expect(optionSelect).toHaveTextContent(/Create new workout/i);
  screen.debug();
});
