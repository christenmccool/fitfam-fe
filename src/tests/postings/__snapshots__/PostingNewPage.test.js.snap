// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    align="center"
    class="MuiContainer-root MuiContainer-maxWidthMd css-1lzhms0-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-8a3xgi"
    >
      <div
        class="MuiBox-root css-134w9h5"
      >
        <div
          class="MuiBox-root css-hpgf8j"
        >
          <div
            class="MuiBox-root css-1l4w6pd"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1tazx64-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="mui-3"
                id="mui-3-label"
              >
                Date
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="Choose date, selected date is Mar 25, 2022"
                  aria-readonly="true"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-3"
                  readonly=""
                  type="text"
                  value="03/25/2022"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-7nxzgd"
                  >
                    <span>
                      Date
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-hpgf8j"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-fuzc6w-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="primFamId"
                id="primFamId-label"
              >
                Select FitFam
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                style="font-size: 20px;"
                variant="outlined"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="primFamId-label primFamId"
                  class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="primFamId"
                  role="button"
                  tabindex="0"
                >
                  Family 1
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="1"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-7nxzgd"
                  >
                    <span>
                      Select FitFam
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-zboufa-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="option"
            id="option-label"
          >
            Select option
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 20px; color: rgb(63, 80, 181);"
            variant="outlined"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="option-label option"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="option"
              role="button"
              tabindex="0"
            >
              Create new workout
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="create"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-7nxzgd"
              >
                <span>
                  Select option
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-178yklu"
      >
        <form
          class="MuiBox-root css-164r41r"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="woName"
                  id="woName-label"
                >
                  Workout Name
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="woName"
                    name="woName"
                    style="font-size: 20px; text-align: center; line-height: normal;"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1g8cwq9"
                    >
                      <span>
                        Workout Name
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="woDescription"
                  id="woDescription-label"
                >
                  Workout Description
                  <span
                    aria-hidden="true"
                    class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id="woDescription"
                    name="woDescription"
                    required=""
                    style="font-size: 20px; text-align: center; line-height: normal; height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); font-size: 20px; text-align: center; line-height: normal; padding: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1g8cwq9"
                    >
                      <span>
                        Workout Description *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1n3sxlg-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-jh47zj-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="submit"
              >
                Submit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-16g5rpk-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButtonBase-root  css-kpm0tx-MuiButtonBase-root-MuiButton-root"
            href="/?date=2022-03-25"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
            >
              Cancel
            </p>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
